File: nudge-next/src/features/ai-graph/guards.ts
/**
 * Safety guards for the universal AI graph — a 1:1 port of
 * `backend/ai_graph/guards.py`.
 *
 * Keep the AI from doing destructive or nonsensical things:
 *   - isOptOut          — robust opt-out / "don't message me" detection
 *   - isMeaningfulText  — empty / whitespace-only guard
 *   - canAiUpdateStage  — pipeline transition guard (no regression, no
 *                         overriding human-owned or terminal states)
 *
 * Pure functions, no DB — easy to unit-test.
 */
import type { Intent, Pack } from "./types";

/** Completed / closed states the AI must never move a lead OUT of. */
export const HARD_TERMINAL: ReadonlySet<string> = new Set([
  "won", "enrolled", "converted", "closed", "booked", "ordered",
  "po_received", "travelled", "served", "fulfilled", "consulted", "visited",
  "resolved",
]);

/** 'lost' is a soft-terminal: only a fresh buying intent may revive it. */
export const REVIVE_INTENTS: ReadonlySet<string> = new Set([
  "PRICE_QUERY", "BOOKING_QUERY", "NEW_LEAD", "PAYMENT_QUERY", "RESCHEDULE",
]);

/**
 * Opt-out: short ambiguous words must match the whole message; clear phrases
 * match as substrings.
 */
const _OPTOUT_EXACT: ReadonlySet<string> = new Set([
  "stop", "cancel", "unsubscribe", "optout", "opt out", "remove me", "stop promotions",
]);
const _OPTOUT_PHRASES: readonly string[] = [
  "unsubscribe", "opt out", "opt-out", "stop promotions", "stop messaging",
  "don't message", "do not message", "dont message", "remove me from", "leave me alone",
];

export function isOptOut(text: string | null | undefined): boolean {
  const t = (text ?? "").trim().toLowerCase();
  if (!t) return false;
  if (_OPTOUT_EXACT.has(t)) return true;
  return _OPTOUT_PHRASES.some((p) => t.includes(p));
}

/** True if there's actual text to reason about (not empty/whitespace). */
export function isMeaningfulText(text: string | null | undefined): boolean {
  return Boolean((text ?? "").trim());
}

/**
 * Decide whether the AI may move a lead from `current` to `proposed`.
 *
 * Blocks: no-ops, human-owned conversations, moving out of terminal states,
 * reviving 'lost' without a fresh buying intent, and stage regressions.
 */
export function canAiUpdateStage(
  current: string | null | undefined,
  proposed: string | null | undefined,
  pack: Pack,
  opts: { humanOwned: boolean; intent: Intent | string | null | undefined },
): boolean {
  const { humanOwned, intent } = opts;
  if (!proposed || proposed === current) return false;
  if (humanOwned) return false; // a human owns this lead — AI must not override

  const cur = (current ?? "").toLowerCase();
  if (HARD_TERMINAL.has(cur)) return false; // never move out of won/enrolled/converted/...
  if (cur === "lost") {
    return REVIVE_INTENTS.has((intent ?? "").toString().toUpperCase()); // only revive on new intent
  }

  const stages = pack.pipeline_stages ?? [];
  if (
    current != null &&
    stages.includes(current) &&
    stages.includes(proposed) &&
    stages.indexOf(proposed) < stages.indexOf(current)
  ) {
    return false; // no backward regression
  }

  return true;
}

File: backend/ai_graph/guards.py
"""
Safety guards for the universal AI graph.

These keep the AI from doing destructive or nonsensical things when real users
send weird inputs or when humans are already involved:

  * is_opt_out          — robust opt-out / "don't message me" detection
  * is_meaningful_text  — empty / whitespace-only guard
  * can_ai_update_stage — pipeline transition guard (no regression, no
                          overriding human-owned or terminal states)

Pure functions, no DB — easy to unit-test.
"""
from __future__ import annotations

# Completed / closed states the AI must never move a lead OUT of.
HARD_TERMINAL = {
    "won", "enrolled", "converted", "closed", "booked", "ordered",
    "po_received", "travelled", "served", "fulfilled", "consulted", "visited",
    "resolved",
}

# 'lost' is a soft-terminal: only a fresh buying intent may revive it.
REVIVE_INTENTS = {"PRICE_QUERY", "BOOKING_QUERY", "NEW_LEAD", "PAYMENT_QUERY", "RESCHEDULE"}

# Opt-out: short ambiguous words must match the whole message; clear phrases
# match as substrings.
_OPTOUT_EXACT = {"stop", "cancel", "unsubscribe", "optout", "opt out", "remove me", "stop promotions"}
_OPTOUT_PHRASES = [
    "unsubscribe", "opt out", "opt-out", "stop promotions", "stop messaging",
    "don't message", "do not message", "dont message", "remove me from", "leave me alone",
]


def is_opt_out(text: str | None) -> bool:
    t = (text or "").strip().lower()
    if not t:
        return False
    if t in _OPTOUT_EXACT:
        return True
    return any(p in t for p in _OPTOUT_PHRASES)


def is_meaningful_text(text: str | None) -> bool:
    """True if there's actual text to reason about (not empty/whitespace)."""
    return bool((text or "").strip())


def can_ai_update_stage(current: str | None, proposed: str | None, pack: dict,
                        *, human_owned: bool, intent: str | None) -> bool:
    """Decide whether the AI may move a lead from `current` to `proposed`.

    Blocks: no-ops, human-owned conversations, moving out of terminal states,
    reviving 'lost' without a fresh buying intent, and stage regressions.
    """
    if not proposed or proposed == current:
        return False
    if human_owned:
        return False  # a human owns this lead — AI must not override

    cur = (current or "").lower()
    if cur in HARD_TERMINAL:
        return False  # never move out of won/enrolled/converted/...
    if cur == "lost":
        return (intent or "").upper() in REVIVE_INTENTS  # only revive on new intent

    stages = pack.get("pipeline_stages", [])
    if current in stages and proposed in stages and stages.index(proposed) < stages.index(current):
        return False  # no backward regression

    return True
